File: AtomykPlay/app/src/main/java/com/atomykcoder/atomykplay/fragments/LastAddedFragment.java
package com.atomykcoder.atomykplay.fragments;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.app.ProgressDialog;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.atomykcoder.atomykplay.R;
import com.atomykcoder.atomykplay.activities.MainActivity;
import com.atomykcoder.atomykplay.adapters.MusicMainAdapter;
import com.atomykcoder.atomykplay.helperFunctions.StorageUtil;
import com.atomykcoder.atomykplay.viewModals.MusicDataCapsule;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.Locale;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;


public class LastAddedFragment extends Fragment {
    private final int firstOptionValue = 30;
    private final int secondOptionValue = 90;
    private final int thirdOptionValue = 180;
    private final int fourthOptionValue = 360;
    private RadioButton firstRadio, secondRadio, thirdRadio, fourthRadio;
    private MusicMainAdapter adapter;
    private ArrayList<MusicDataCapsule> lastAddedMusicList;
    private Dialog filterDialog;
    private TextView songCountTv;
    private StorageUtil.SettingsStorage settingsStorage;
    private ArrayList<MusicDataCapsule> initialMusicList;
    private ProgressDialog progressDialog;
    private RecyclerView recyclerView;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_last_added, container, false);

        //get references
        recyclerView = view.findViewById(R.id.last_added_recycle_view);
        settingsStorage = new StorageUtil.SettingsStorage(requireContext());
        View filterButton = view.findViewById(R.id.filter_last_added_btn);
        ImageView backImageView = view.findViewById(R.id.close_filter_btn);
        progressDialog = new ProgressDialog(requireContext());
        songCountTv = view.findViewById(R.id.count_of_lastAdded);
        // initialize/load music array lists
        initialMusicList = new StorageUtil(getContext()).loadInitialList();
        lastAddedMusicList = new ArrayList<>();



        //sort initial music list by date in reverse order
        Collections.sort(initialMusicList, new Comparator<MusicDataCapsule>() {
            final DateFormat dateFormat = new SimpleDateFormat("dd MMMM yyyy", Locale.getDefault());
            @Override
            public int compare(MusicDataCapsule t1, MusicDataCapsule t2) {
                try {
                    Date d1 = dateFormat.parse(t1.getsDateAdded());
                    Date d2 = dateFormat.parse(t2.getsDateAdded());
                    int i = 0;
                    if (d1 != null) {
                        i = d1.compareTo(d2);
                    }
                    if(i != 0) return -i;
                } catch (ParseException e) {
                    throw new IllegalArgumentException();
                }
                return 0;
            }
        });

        loadLastAddedList(settingsStorage.loadLastAddedDur());

        // back button click listener
        backImageView.setOnClickListener(v -> requireActivity().onBackPressed());

        // filter click listener
        filterButton.setOnClickListener(v -> openDialogFilter());

        return view;
    }

    @Override
    public void onStart() {
        super.onStart();
    }

    @SuppressLint("SetTextI18n")
    private void openDialogFilter() {
        filterDialog = new Dialog(requireContext());
        filterDialog.setContentView(R.layout.last_added_filter_dialog);

        int savedState = settingsStorage.loadLastAddedDur();

        RadioGroup radioGroup1 = filterDialog.findViewById(R.id.last_Added_radio_group);
        firstRadio = filterDialog.findViewById(R.id.last_added_rb_1);
        secondRadio = filterDialog.findViewById(R.id.last_added_rb_2);
        thirdRadio = filterDialog.findViewById(R.id.last_added_rb_3);
        fourthRadio = filterDialog.findViewById(R.id.last_added_rb_4);

        switch (savedState) {
            case 1:
                firstRadio.setChecked(true);
                break;
            case 2:
                secondRadio.setChecked(true);
                break;
            case 3:
                thirdRadio.setChecked(true);
                break;
            case 4:
                fourthRadio.setChecked(true);
                break;
        }

        firstRadio.setText("Last month");
        secondRadio.setText("Last three months");
        thirdRadio.setText("Last six months");
        fourthRadio.setText("Last year");

        radioGroup1.setOnCheckedChangeListener((radioGroup, in) -> {
            int i = getRadioID();
            loadLastAddedList(i);
            filterDialog.dismiss();
        });
        filterDialog.show();
    }



    @Override
    public void onStop() {
        super.onStop();
        if (filterDialog != null) {
            if (filterDialog.isShowing()) {
                filterDialog.dismiss();
            }
        }
    }



    /**
     * load all last added songs based on selected radio buttons
     * @param i radio id
     */
    private void loadLastAddedList(int i) {
        switch (i) {
            case 1:
                startThread(firstOptionValue);
                break;
            case 2:
                startThread(secondOptionValue);
                break;
            case 3:
                startThread(thirdOptionValue);
                break;
            case 4:
                startThread(fourthOptionValue);
                break;
        }
        settingsStorage.setLastAddedDur(i);
    }

    /**
     * start thread to handle loading music list within range
     * @param maxValue clamp at max value
     */
    private void startThread(int maxValue) {
        ExecutorService service = Executors.newSingleThreadExecutor();
        Handler handler = new Handler(Looper.getMainLooper());

        progressDialog.setMessage("Calculating...");
        progressDialog.show();

        service.execute(() -> {

                lastAddedMusicList.clear();
            try {
                lastAddedMusicList.addAll(getLastAddedMusicList(maxValue));
            } catch (ParseException e) {
                e.printStackTrace();
            }

            handler.post(() -> {

                progressDialog.dismiss();
                String num = lastAddedMusicList.size() + " Songs";
                songCountTv.setText(num);

                // set recyclerview and adapter
                recyclerView.setHasFixedSize(true);
                recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
                ArrayList<String> idList = new ArrayList<>();
                for (MusicDataCapsule music : lastAddedMusicList) {
                    idList.add(music.getsId());
                }
                adapter = new MusicMainAdapter(getContext(), lastAddedMusicList,idList);
                recyclerView.setAdapter(adapter);
                });
        });
        service.shutdown();
    }


    private ArrayList<MusicDataCapsule> getLastAddedMusicList (long max) throws ParseException {
        ArrayList<MusicDataCapsule> result = new ArrayList<>();

        // dtf for parsing string to date
        DateFormat dateFormat = new SimpleDateFormat("dd MMMM yyyy", Locale.getDefault());

        //get "current" date
        Date previousDate = dateFormat.parse(dateFormat.format(new Date()));

        // subtract days from current date and turn it into an older date
        if (previousDate != null) {
            previousDate.setTime(previousDate.getTime() - (max * 86400) * 1000);
        }

        //loop through music list and if a music date is older than given date, then break;
        for (MusicDataCapsule music : initialMusicList) {
            Date musicDate = dateFormat.parse(music.getsDateAdded());
            if(musicDate.compareTo(previousDate) < 0) {
                break;
            }
                result.add(music);
        }

        //return music ranging between current date and given older date
        return result;
    }


    /**
     * radio id check
     *
     * @return returns selected radio button id
     */
    private int getRadioID() {
        int radioId = 0;

        if (firstRadio.isChecked()) {
            radioId = 1;
        } else if (secondRadio.isChecked()) {
            radioId = 2;
        } else if (thirdRadio.isChecked()) {
            radioId = 3;
        } else if (fourthRadio.isChecked()) {
            radioId = 4;
        }
        return radioId;
    }

    @SuppressLint("NotifyDataSetChanged")
    private void notifyAdapter() {
        adapter.notifyDataSetChanged();
    }

}

File: AtomykPlay/app/src/main/java/com/atomykcoder/atomykplay/fragments/LastAddedFragment.kt
package com.atomykcoder.atomykplay.fragments

import android.annotation.SuppressLint
import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import com.atomykcoder.atomykplay.R
import com.atomykcoder.atomykplay.adapters.MusicMainAdapter
import com.atomykcoder.atomykplay.data.Music
import com.atomykcoder.atomykplay.scripts.LinearLayoutManagerWrapper
import com.atomykcoder.atomykplay.utils.StorageUtil
import com.atomykcoder.atomykplay.utils.StorageUtil.SettingsStorage
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import kotlinx.coroutines.*
import java.text.DateFormat
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

class LastAddedFragment : Fragment() {
    companion object{
        @JvmStatic
        fun newInstance() = LastAddedFragment()
    }
    private lateinit var recyclerView: RecyclerView
    private val firstOptionValue = 30
    private val secondOptionValue = 90
    private val thirdOptionValue = 180
    private val fourthOptionValue = 360
    private var adapter: MusicMainAdapter? = null
    private var lastAddedMusicList: ArrayList<Music>? = null
    private var filterDialog: Dialog? = null
    private var songCountTv: TextView? = null
    private var settingsStorage: SettingsStorage? = null
    private var initialMusicList: ArrayList<Music>? = null
    private var progressDialog: ProgressBar? = null
    private val coroutineMainScope = CoroutineScope(Dispatchers.Main)
    private val coroutineDefaultScope = CoroutineScope(Dispatchers.Default)

    fun removeItems(selectedItem: Music?) {
        selectedItem?.let { adapter?.removeItem(it) }
    }


    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View? {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_last_added, container, false)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        //get references

        recyclerView = view.findViewById(R.id.last_added_recycle_view)
        //set recyclerview and adapter
        recyclerView.setHasFixedSize(true)
        recyclerView.setItemViewCacheSize(5)
        recyclerView.layoutManager = LinearLayoutManagerWrapper(context)
        settingsStorage = SettingsStorage(requireContext())
        val filterButton = view.findViewById<View>(R.id.filter_last_added_btn)
        val backImageView = view.findViewById<ImageView>(R.id.close_filter_btn)
        progressDialog = view.findViewById(R.id.progress_bar_last_added)
        songCountTv = view.findViewById(R.id.count_of_lastAdded)
        // initialize/load music array lists
        initialMusicList = StorageUtil(requireContext()).loadInitialList()
        lastAddedMusicList = ArrayList()

        // back button click listener
        backImageView.setOnClickListener { requireActivity().onBackPressed() }

        // filter click listener
        filterButton.setOnClickListener { openDialogFilter() }

        coroutineDefaultScope.launch {
            //sort initial music list by date in reverse order
            val locale: Locale = Locale.getDefault()
            initialMusicList?.sortWith(object : Comparator<Music> {
                val dateFormat: DateFormat = SimpleDateFormat("dd MMMM yyyy", locale)
                override fun compare(t1: Music, t2: Music): Int {
                    try {
                        val d1 = dateFormat.parse(t1.dateAdded)
                        val d2 = dateFormat.parse(t2.dateAdded)
                        var i = 0
                        if (d1 != null) {
                            i = d1.compareTo(d2)
                        }
                        if (i != 0) return -i
                    } catch (e: ParseException) {
                        e.printStackTrace()
                    }
                    return 0
                }
            })
            lastAddedMusicList = initialMusicList?.let { ArrayList(it) }
            coroutineMainScope.launch {
                adapter = MusicMainAdapter(requireContext(), lastAddedMusicList)
                recyclerView.adapter = adapter
                loadLastAddedList(settingsStorage?.loadLastAddedDur())
            }
        }
    }

    override fun onStop() {
        super.onStop()
        if (filterDialog != null) {
            if (filterDialog!!.isShowing) {
                filterDialog!!.dismiss()
            }
        }
    }

    @SuppressLint("SetTextI18n")
    private fun openDialogFilter() {
        val builder = MaterialAlertDialogBuilder(requireContext())
        builder.setTitle("Filter")
        val savedState = settingsStorage?.loadLastAddedDur() ?: 0
        val options = arrayOf("Last month", "Last three months", "Last six months", "Last year")
        builder.setSingleChoiceItems(options, savedState) { dialog, which ->
            loadLastAddedList(which)
            dialog.dismiss()
        }

        filterDialog = builder.create()
        filterDialog?.show()
    }

    /**
     * load all last added songs based on selected radio buttons
     *
     * @param i radio id
     */
    private fun loadLastAddedList(i: Int?) {
        when (i) {
            0 -> startThread(firstOptionValue)
            1 -> startThread(secondOptionValue)
            2 -> startThread(thirdOptionValue)
            3 -> startThread(fourthOptionValue)
        }
        settingsStorage!!.setLastAddedDur(i ?: 0)
    }

    /**
     * start thread to handle loading music list within range
     *
     * @param maxValue clamp at max value
     */
    private fun startThread(maxValue: Int) {
        progressDialog!!.visibility = View.VISIBLE
        coroutineDefaultScope.launch {
            lastAddedMusicList!!.clear()
            try {
                lastAddedMusicList!!.addAll(getLastAddedMusicList(maxValue.toLong()))
            } catch (e: ParseException) {
                e.printStackTrace()
            }
            coroutineMainScope.launch {
                progressDialog!!.visibility = View.GONE
                val num = lastAddedMusicList?.size?.toString() + " Songs"
                songCountTv!!.text = num
                notifyAdapter()
            }
        }
    }

    @Throws(ParseException::class)
    private fun getLastAddedMusicList(max: Long): ArrayList<Music> {
        val result = ArrayList<Music>()

        // dtf for parsing string to date
        val dateFormat: DateFormat = SimpleDateFormat("dd MMMM yyyy", Locale.getDefault())

        //get "current" date
        val previousDate = dateFormat.parse(dateFormat.format(Date()))

        // subtract days from current date and turn it into an older date
        if (previousDate != null) {
            previousDate.time = previousDate.time - max * 86400 * 1000
        }

        //loop through music list and if a music date is older than given date, then break;
        for (music in initialMusicList!!) {
            val musicDate = dateFormat.parse(music.dateAdded)
            if (musicDate != null && musicDate < previousDate) {
                break
            }
            result.add(music)
        }

        //return music ranging between current date and given older date
        return result
    }

    @SuppressLint("NotifyDataSetChanged")
    private fun notifyAdapter() {
        adapter!!.notifyDataSetChanged()
    }
}